### PR TITLE
Identify statically-linked symbols that are needed by modules

### PR DESCRIPTION
**Pull request by bonzini**
Even though shared modules are more or less incompatible with statically linked binaries (glibc's support for libdl inside statically linked binaries, for example, is very limited), it is of course possible to link static libraries into a binary that uses shared modules. These could be either external packages, or libraries that are internal to the program being built.
    
Whenever an executable is statically linked with a library, the linker omits library functions that are not used by the executable.  However, some of these omitted library functions, while not used by the executable, might be needed by shared modules.  When this happens, the shared module will fail to load.

Because the library is statically linked, adding it to the module may not do the right thing. The most common workaround in that case is to use link_whole, but that might seriously bloat the executable and might not even always work. This pull request implements an alternative solution.  It is inspired by the operation of QEMU's Makefiles, which I'm looking at for a conversion to Meson.

Module object files are processed with "nm" to find undefined symbols, and the undefined symbols are included in the executable through the linker's "-u" option.  There exists equivalents in the MSVC world for both steps, but they are not implemented yet. The process is automatic and guided entirely by syntax that is already present in the Meson language.
    
Note that, unlike symbolextractor, this is not just an optimization. However, if for example nm is not found in the toolchain file for cross-compilation, behavior is unchanged compared to before this pull request: a module that has dependencies on a statically-linked library will fail to link, but apart from this special case everything will work normally.

### CODE REVIEW

File: mesonbuild/scripts/undefsymbols.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script extracts undefined symbols of a given library
+# into a file. If the symbols have not changed, the file is not
+# touched. This information is used to include functions from static
+# libraries into the main program.
+
+import os, sys
+from .. import mesonlib
```
Comment (ignatenkobrain) [inline mesonbuild/scripts/undefsymbols.py]: __[Flake8]__

[F401] '..mesonlib' imported but unused

File: mesonbuild/scripts/undefsymbols.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script extracts undefined symbols of a given library
+# into a file. If the symbols have not changed, the file is not
+# touched. This information is used to include functions from static
+# libraries into the main program.
+
+import os, sys
```
Comment (ignatenkobrain) [inline mesonbuild/scripts/undefsymbols.py]: __[Flake8]__

[F401] 'os' imported but unused

File: mesonbuild/backend/ninjabackend.py [hunk 3]
```diff
@@ -1332,7 +1326,9 @@ def generate_rust_target(self, target, outfile):
         element.add_item('cratetype', cratetype)
         element.write(outfile)
         if isinstance(target, build.SharedLibrary):
-            self.generate_shsym(outfile, target)
+            self.generate_shsym(rust + crstr, outfile, target)
+        if isinstance(target, build.SharedModule):
+            self.generate_undefsym(rust + crst, outfile, target)
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[Flake8]__

[F821] undefined name 'rust'

File: mesonbuild/backend/ninjabackend.py [hunk 4]
```diff
@@ -1332,7 +1326,9 @@ def generate_rust_target(self, target, outfile):
         element.add_item('cratetype', cratetype)
         element.write(outfile)
         if isinstance(target, build.SharedLibrary):
-            self.generate_shsym(outfile, target)
+            self.generate_shsym(rust + crstr, outfile, target)
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[Flake8]__

[F821] undefined name 'rust'

File: mesonbuild/backend/ninjabackend.py [hunk 5]
```diff
@@ -1337,7 +1331,9 @@ def generate_rust_target(self, target, outfile):
         element.add_item('cratetype', cratetype)
         element.write(outfile)
         if isinstance(target, build.SharedLibrary):
-            self.generate_shsym(outfile, target)
+            self.generate_shsym('rust' + crstr, outfile, target)
+        if isinstance(target, build.SharedModule):
+            self.generate_undefsym('rust' + crst, outfile, target)
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[Flake8]__

[F821] undefined name 'crst'

File: mesonbuild/compilers/compilers.py [hunk 6]
```diff
@@ -49,7 +49,7 @@
 c_suffixes = lang_suffixes['c'] + ('h',)
 # List of languages that can be linked with C code directly by the linker
 # used in build.py:process_compilers() and build.py:get_dynamic_linker()
-clike_langs = ('d', 'objcpp', 'cpp', 'objc', 'c', 'fortran',)
+clike_langs = ('d', 'objcpp', 'cpp', 'objc', 'c', 'fortran', 'vala',)
```
Comment (nirbheek) [inline mesonbuild/compilers/compilers.py]: Please open a separate PR for this, it's bad practice to mix commits from unrelated changes together. It also means that your work will get pushed faster.

Please include the sentence "Fixes https://github.com/mesonbuild/meson/issues/791" in your commit message so the issue is closed automatically.

Comment (bonzini) [reply mesonbuild/compilers/compilers.py]: Ok, I was planning to do so but I wasn't sure how to get CI to work correctly on this PR. Thanks!

**Comment by bonzini** [comment]
I think I will reimplement this after adding the above feature to Ninja. However the main ideas (scripts, rules) will be the same.

**Comment by bonzini** [comment]
> Do you have an estimate for how much bloat this adds?

For one of the binaries built by QEMU, which only links with glib (libglib-2.0.a and libgthread-2.0.a) and glibc:

    qemu-x86_64 with statically linked glib: text + data = 5946217
    qemu-x86_64 with whole-archive glib: text + data = 6709897

So the extra size is about 12% of the binary. The space taken by glib itself in the binary grows by a factor of 2, from ~630 KiB to ~1380 KiB.

> > and might not even always work
>
> When would this happen?

In the case of QEMU, there is a "utility functions" static library used by both the emulators and the other smaller programs distributed with QEMU. Whenever some functionality is only required in the emulators, the static library includes stubs for use in the smaller tools, while the emulators override the stubs. whole_archive would cause duplicate definitions in the emulators, and linking would fail.

In the past we tried other mechanisms such as weak symbols, but static libraries turned out to be the easiest and most portable way to implement this kind of stubbing.

**Comment by bonzini** [comment]
Looks like all CI tests pass now. Sorry for the noise!

**Comment by jpakkane** [comment]
Am I correct in reading the code that this is to adding a link argument like `-Wl,-u,foofunc` for all the missing functions?

I'm concerned about adding something automatic like this for what is a fairly niche use case. One of the things we've learned that changes like these _always_ break someone.

**Comment by bonzini** [comment]
> Am I correct in reading the code that this is to adding a link argument like `-Wl,-u,foofunc` for all the missing functions?

Yes. Note however that `-u` is just a hint. Compare these two:

    $ cat f.c
    int main() { }
    $ gcc f.c -Wl,-u,bar
    $ gcc f.c -Wl,--require-defined=bar
    /usr/bin/ld: required symbol `bar' not defined
    collect2: error: ld returned 1 exit status

What do you suggest? A built-in option?

**Comment by jpakkane** [comment]
Well, AIUI this can be done without these changes with something like this:

```meson
required_symbols_args = ['-Wl,--require-defined=bar', ...]

executable(...,
     link_args: required_symbol_args)
```

The downside being that you don't get the list of symbols automatically. On the other hand listing them out is more explicit which some people consider to be an advantage. I guess it depends how often the list of symbols changes...

**Comment by bonzini** [comment]
> I guess it depends how often the list of symbols changes...

It can change quite often. With our current build system it's totally transparent so I don't know how often it would change, but it would likely be a blocker for switching to Meson. We never had a compatibility problem.

I guess it would be okay to keep the logic in the build system instead of moving it to Meson. There are two issues for this:

1. VC++ does not support nested response files so it would not be possible to support it. But my project does not care so this is not a blocker.

2. You cannot order a `custom_target` before an `executable`. For example:

    project('a', 'c')
    lib_util = static_library('util', 'x.c', pic: true)
    mod_b = shared_module('b', 'y.c')
    target = custom_target('undefsym', output: ['x.undef'], input: mod_b,
                           command: ['undefsym.sh', '@INPUT@'])
    lib_util_full = declare_dependency(link_with: [lib_util], link_args: '@x.undef')
    executable('a', 'f.c', dependencies: [lib_util_full])

but there is no way for `lib_util_full` to add an implicit dependency from the executable to `x.undef`. If you can suggest how to extend `declare_dependency` I can work on it.

**Comment by jpakkane** [comment]
Well one way of doing this currently, but which granted is a bit hacky is to do something like:

```meson
# UNDEF_DEFINITION START
<add here all the link args needed>
# UNDEF_DEFINITION END

run_target(# a script that takes the files as input, opens this file and updates the linker 
                  # command list above
)
```

This would not be automatic (you'd need to execute the run target manually when symbols change) but it would work for all backends and cases.

**Comment by bonzini** [comment]
The set of symbols depends on the target (e.g. Linux/Windows/Mac) and especially for libraries such as glib it's not very predictable when it changes. The same patch, backported to a stable branch, might necessitate adding a symbol or two.

I guess CI would catch mistakes but still I don't think it would be acceptable for other developers.

The only solution I see is being able to add implicit dependencies to the link rule via `declare_dependency`.

**Comment by bonzini** [comment]
(Of course the manual solution was already considered and discarded when we came up with the current scheme using nm).

**Comment by bonzini** [comment]
Ok, I got it to work. The trick is: 1) to build the modules in two phases using `link_whole`, so that the list of symbols can be gleaned from the modules' objects file, but the executable can still be linked before the shared module; 2) to use link_depends so that the symbols are updated automatically:

    project('a', 'c')
    
    modules_objs = []
    modules_objs += [static_library('b.module', 'y.c', pic: true)]
    # ...

    target = custom_target('undefsym', output: ['x.undef'], input: modules_objs,
                           command: [files('undefsym.sh'), '@OUTPUT0@', '@INPUT@'])
    
    libutil = static_library('util', 'x.c', pic: true)
    e = executable('a', 'f.c', link_with: [libutil], link_args: '@x.undef', link_depends: target, implib: true)
    foreach m : modules_objs
        shared_module('b', link_whole: m, link_with: e)
    endforeach

For the full solution, I would use `get_cross_property` to pass the `nm` binary to the script. It's not great, but I guess it's acceptable. The only extra build step compared to my patches is the `ar` to build `libb.module.a`.

**Comment by jpakkane** [comment]
I actually woke up last night and then immediately realized that `link_depends` can be used to make this work. This is not a dramatization, it really did happen. Good that you realized the same thing.

>  For the full solution, I would use `get_cross_property` to pass the `nm` binary to the script. It's not great, but I guess it's acceptable.

If you have set up nm in the cross file (which you should) then doing `nm_exe = find_program('nm')` returns the cross compilation nm rather than the native one.

**Comment by elmarco** [comment]
I managed to build the join of undefs & defs as a custom target, but I wonder what tricks I could use to transform the result (content of target file) as link_args. I wish I could have a custom_target().stdout() for example. (similar to run_command() run result object). Any idea?

**Comment by bonzini** [comment]
You can use `@foo` to include the content of a file into the command line (this is a gcc feature that is also supported by clang).
